File: app/src/main/java/com/esorokin/lantector/ui/plugins/base/CompositionPluginDelegate.java
package com.esorokin.lantector.ui.plugins.base;

import java.util.ArrayList;
import java.util.Collection;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.view.MenuItem;
import android.view.View;

public final class CompositionPluginDelegate implements CompositionPlugin {
	public Collection<Plugin> plugins = new ArrayList<>();

	@Override
	public void attach(final Plugin plugin) {
		if (plugin != null) {
			plugins.add(plugin);
		}
	}

	@Override
	public void detach(final Plugin plugin) {
		if (plugin != null) {
			plugins.remove(plugin);
		}
	}

	@Override
	public void onCreate() {
		for (Plugin plugin : plugins) {
			plugin.onCreate();
		}
	}

	@Override
	public void onViewCreated(final View view) {
		for (Plugin plugin : plugins) {
			plugin.onViewCreated(view);
		}
	}

	@Override
	public void onStart() {
		for (Plugin plugin : plugins) {
			plugin.onStart();
		}
	}

	@Override
	public void onResume() {
		for (Plugin plugin : plugins) {
			plugin.onResume();
		}
	}

	@Override
	public void onPause() {
		for (Plugin plugin : plugins) {
			plugin.onPause();
		}
	}

	@Override
	public void onStop() {
		for (Plugin plugin : plugins) {
			plugin.onStop();
		}
	}

	@Override
	public void onDestroy() {
		for (Plugin plugin : plugins) {
			plugin.onDestroy();
		}
	}

	@Override
	public void onOptionsItemSelected(MenuItem item) {
		for (Plugin plugin : plugins) {
			plugin.onOptionsItemSelected(item);
		}
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
		for (Plugin plugin : plugins) {
			plugin.onRequestPermissionsResult(requestCode, permissions, grantResults);
		}
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data) {
		for (Plugin plugin : plugins) {
			plugin.onActivityResult(requestCode, resultCode, data);
		}
	}

	@Override
	public void onBackPressed() {
		for (Plugin plugin : plugins) {
			plugin.onBackPressed();
		}
	}
}

File: app/src/main/java/com/esorokin/lantector/ui/plugin/base/CompositionPluginDelegate.kt
package com.esorokin.lantector.ui.plugin.base

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View

class CompositionPluginDelegate : CompositionPlugin {
    var plugins: MutableList<Plugin> = mutableListOf()

    override fun <T : Plugin> attach(plugin: T): T = plugin.also { plugins.add(it) }

    override fun detach(plugin: Plugin) {
        plugins.remove(plugin)
    }

    override fun onCreate(savedInstanceState: Bundle?) = plugins.forEach { it.onCreate(savedInstanceState) }

    override fun onViewCreated(view: View) = plugins.forEach { it.onViewCreated(view) }

    override fun onStart() = plugins.forEach { it.onStart() }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) = plugins.forEach { it.onRestoreInstanceState(savedInstanceState) }

    override fun onResume() = plugins.forEach { it.onResume() }

    override fun onPause() = plugins.forEach { it.onPause() }

    override fun onSaveInstanceState(outState: Bundle?) = plugins.forEach { it.onSaveInstanceState(outState) }

    override fun onStop() = plugins.forEach { it.onStop() }

    override fun onDestroy() = plugins.forEach { it.onDestroy() }

    override fun onOptionsItemSelected(item: MenuItem) = plugins.forEach { it.onOptionsItemSelected(item) }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) = plugins.forEach { it.onRequestPermissionsResult(requestCode, permissions, grantResults) }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) = plugins.forEach { it.onActivityResult(requestCode, resultCode, data) }

    override fun onBackPressed() = plugins.forEach { it.onBackPressed() }
}